bundles-lib/src/main/java/org/apache/metron/bundles/ExtensionManager.java

```diff
@@ -0,0 +1,436 @@
+   * Bundle library path, which will become the working directory of the returned bundle
+   * @return a bundle for the system class loader
+   */
+  public static Bundle createSystemBundle(final FileSystemManager fileSystemManager,
+      final BundleProperties bundleProperties) throws FileSystemException, URISyntaxException {
+    final ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
+
+    final String bundlesLibraryDirectory = bundleProperties
+        .getProperty(BundleProperties.BUNDLE_LIBRARY_DIRECTORY);
+    if (StringUtils.isBlank(bundlesLibraryDirectory)) {
+      throw new IllegalStateException(
+          "Unable to create system bundle because " + BundleProperties.BUNDLE_LIBRARY_DIRECTORY
+              + " was null or empty");
+    }
+    final URI bundlesLibraryDirURI = bundleProperties.getBundleLibraryDirectory();
+    FileObject bundleDir = fileSystemManager.resolveFile(bundlesLibraryDirURI);
+
+    // Test if the source Bundles can be read
+    FileUtils.ensureDirectoryExistAndCanRead(bundleDir);
+
+    final BundleDetails systemBundleDetails = new BundleDetails.Builder()
+        .withBundleFile(bundleDir)
+        .withCoordinates(SYSTEM_BUNDLE_COORDINATE)
+        .build();
+
+    return new Bundle(systemBundleDetails, systemClassLoader);
+  }
+
+  /**
+   * Loads extensions from the specified bundle.
+   *
+   * @param bundle from which to load extensions
+   */
+  @SuppressWarnings("unchecked")
+  private static void loadExtensions(final Bundle bundle) throws NotInitializedException {
+    checkInitialized();
+    for (final Map.Entry<Class, Set<Class>> entry : definitionMap.entrySet()) {
+      // this is another extention point
+      // what we care about here is getting the right classes from the classloader for the bundle
+      // this *could* be as a 'service' itself with different implementations
+      // The NAR system uses the ServiceLoader, but this chokes on abstract classes, because for some
+      // reason it feels compelled to instantiate the class,
+      // which there may be in the system.
+      // Changed to ClassIndex
+      Class clazz = entry.getKey();
+      ClassLoader cl = bundle.getClassLoader();
+      Iterable<Class<?>> it = ClassIndex.getSubclasses(clazz, cl);
+      for (Class<?> c : it) {
+        if (cl.equals(c.getClassLoader())) {
+          // check for abstract
+          if (!Modifier.isAbstract(c.getModifiers())) {
+            registerServiceClass(c, classNameBundleLookup, bundle, entry.getValue());
+          }
+        }
+      }
+      it = ClassIndex.getAnnotated(clazz, cl);
+      for (Class<?> c : it) {
+        if (cl.equals(clazz.getClassLoader())) {
+          // check for abstract
+          if (!Modifier.isAbstract(c.getModifiers())) {
+            registerServiceClass(c, classNameBundleLookup, bundle, entry.getValue());
+          }
+        }
+      }
+
+    }
+  }
+
+
```

Comment:
I believe that the class index process, because it is done at compile/packaging time and not done by HAND as it is in nifi negates the nifi methodology